From 015e0b7592e303e8623b4b46d021ea5bca1ff75c Mon Sep 17 00:00:00 2001
From: "Srivatsa S. Bhat (VMware)" <srivatsa@csail.mit.edu>
Date: Fri, 17 Feb 2023 12:28:37 -0800
Subject: [PATCH] ice: Fix build errors on 6.1.y

src/ice_main.c: In function 'ice_napi_add':
src/ice_main.c:4372:17: error: too many arguments to function 'netif_napi_add'

Signed-off-by: Srivatsa S. Bhat (VMware) <srivatsa@csail.mit.edu>
---
 src/ice_base.c         |  3 +--
 src/ice_eswitch.c      |  3 +--
 src/ice_main.c         |  2 +-
 src/kcompat_impl.h     | 26 ++++++++++++++++++++++++++
 src/kcompat_std_defs.h | 13 +++++++++++++
 5 files changed, 42 insertions(+), 5 deletions(-)

diff --git a/src/ice_base.c b/src/ice_base.c
index b43752e..d9c9d42 100644
--- a/src/ice_base.c
+++ b/src/ice_base.c
@@ -127,8 +127,7 @@ static int ice_vsi_alloc_q_vector(struct ice_vsi *vsi, u16 v_idx)
 	 * handler here (i.e. resume, reset/rebuild, etc.)
 	 */
 	if (vsi->netdev)
-		netif_napi_add(vsi->netdev, &q_vector->napi, ice_napi_poll,
-			       NAPI_POLL_WEIGHT);
+		netif_napi_add(vsi->netdev, &q_vector->napi, ice_napi_poll);
 
 out:
 	/* tie q_vector and VSI together */
diff --git a/src/ice_eswitch.c b/src/ice_eswitch.c
index 6e44820..3c8f65b 100644
--- a/src/ice_eswitch.c
+++ b/src/ice_eswitch.c
@@ -330,8 +330,7 @@ static int ice_eswitch_setup_reprs(struct ice_pf *pf)
 		if (max_vsi_num < vsi->vsi_num)
 			max_vsi_num = vsi->vsi_num;
 
-		netif_napi_add(vf->repr->netdev, &vf->repr->q_vector->napi, ice_napi_poll,
-			       NAPI_POLL_WEIGHT);
+		netif_napi_add(vf->repr->netdev, &vf->repr->q_vector->napi, ice_napi_poll);
 
 		netif_keep_dst(vf->repr->netdev);
 	}
diff --git a/src/ice_main.c b/src/ice_main.c
index 1221d57..235aeb6 100644
--- a/src/ice_main.c
+++ b/src/ice_main.c
@@ -4373,7 +4373,7 @@ static void ice_napi_add(struct ice_vsi *vsi)
 
 	ice_for_each_q_vector(vsi, v_idx)
 		netif_napi_add(vsi->netdev, &vsi->q_vectors[v_idx]->napi,
-			       ice_napi_poll, NAPI_POLL_WEIGHT);
+			       ice_napi_poll);
 }
 
 /**
diff --git a/src/kcompat_impl.h b/src/kcompat_impl.h
index 3212d88..99c53c1 100644
--- a/src/kcompat_impl.h
+++ b/src/kcompat_impl.h
@@ -941,4 +941,30 @@ static inline void eth_hw_addr_set(struct net_device *dev, const u8 *addr)
 #endif /* CONFIG_RETPOLINE */
 #endif /* NEED_EXPORT_INDIRECT_CALLABLE */
 
+/*
+ * NEED_NETIF_NAPI_ADD_NO_WEIGHT
+ *
+ * Upstream commit b48b89f9c189 ("net: drop the weight argument from
+ * netif_napi_add") removes weight argument from function call.
+ *
+ * Our drivers always used default weight, which is 64.
+ *
+ * Define NEED_NETIF_NAPI_ADD_NO_WEIGHT on kernels 3.10+ to use old
+ * implementation. Undef for 6.1+ where new function was introduced.
+ */
+#ifdef NEED_NETIF_NAPI_ADD_NO_WEIGHT
+static inline void
+_kc_netif_napi_add(struct net_device *dev, struct napi_struct *napi,
+		   int (*poll)(struct napi_struct *, int))
+{
+	return netif_napi_add(dev, napi, poll, NAPI_POLL_WEIGHT);
+}
+
+/* RHEL7 complains about redefines. Undef first, then define compat wrapper */
+#ifdef netif_napi_add
+#undef netif_napi_add
+#endif
+#define netif_napi_add _kc_netif_napi_add
+#endif /* NEED_NETIF_NAPI_ADD_NO_WEIGHT */
+
 #endif /* _KCOMPAT_IMPL_H_ */
diff --git a/src/kcompat_std_defs.h b/src/kcompat_std_defs.h
index be000af..51ff393 100644
--- a/src/kcompat_std_defs.h
+++ b/src/kcompat_std_defs.h
@@ -31,6 +31,12 @@
 #error "KERNEL_VERSION is undefined"
 #endif
 
+/*****************************************************************************/
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(3,10,0))
+#else /* >= 3,10,0 */
+#define NEED_NETIF_NAPI_ADD_NO_WEIGHT
+#endif /* 3,10,0 */
+
 /*****************************************************************************/
 #if (LINUX_VERSION_CODE < KERNEL_VERSION(3,18,0))
 #define NEED_DEV_PM_DOMAIN_ATTACH_DETACH
@@ -294,4 +300,11 @@
 #define HAVE_NDO_FDB_DEL_EXTACK
 #endif /* 5.19.0 */
 
+/*****************************************************************************/
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(6,1,0))
+#else /* >=6.1.0 */
+#undef NEED_NETIF_NAPI_ADD_NO_WEIGHT
+#undef HAVE_NDO_GET_DEVLINK_PORT
+#endif /* 6.1.0 */
+
 #endif /* _KCOMPAT_STD_DEFS_H_ */
-- 
2.25.1

